x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredForecastsRemover.java

```diff
@@ -129,39 +125,49 @@ public void onFailure(Exception e) {
         });
     }
 
-    private List<ForecastRequestStats> findForecastsToDelete(SearchResponse searchResponse) throws IOException {
-        List<ForecastRequestStats> forecastsToDelete = new ArrayList<>();
+    private List<JobForecastId> findForecastsToDelete(SearchResponse searchResponse) {
+        List<JobForecastId> forecastsToDelete = new ArrayList<>();
 
         SearchHits hits = searchResponse.getHits();
         if (hits.getTotalHits() > MAX_FORECASTS) {
             LOGGER.info("More than [{}] forecasts were found. This run will only delete [{}] of them", MAX_FORECASTS, MAX_FORECASTS);
         }
 
         for (SearchHit hit : hits.getHits()) {
-            try (InputStream stream = hit.getSourceRef().streamInput();
-                 XContentParser parser = XContentFactory.xContent(XContentType.JSON).createParser(
-                         NamedXContentRegistry.EMPTY, LoggingDeprecationHandler.INSTANCE, stream)) {
-                ForecastRequestStats forecastRequestStats = ForecastRequestStats.LENIENT_PARSER.apply(parser, null);
-                if (forecastRequestStats.getExpiryTime().toEpochMilli() < cutoffEpochMs) {
-                    forecastsToDelete.add(forecastRequestStats);
+            String expiryTime = stringFieldValueOrNull(hit, ForecastRequestStats.EXPIRY_TIME.getPreferredName());
```

Comment:
In 6.8 the doc value could be a `Long` rather than a `String`.  It's why `TimeField` has this case: https://github.com/elastic/elasticsearch/blob/cde202610c58def76e5a923892776c0d7e00f88a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/TimeField.java#L49

What this means in practice is that a user running 6.8 who first used ML in 5.x will end up seeing the warning on lines 139-140 repeatedly and won't get any cleanup.

It's still OK to use `stringFieldValueOrNull()` to extract fields mapped as `keyword` or `text` from hits, but for fields mapped as `date` in 6.8 the code needs to handle both `Long` and `String`.

Reply:
Good point thanks.

I'm curious though why does this code throw if the object is a Long? This is from the 6.8 branch

https://github.com/elastic/elasticsearch/blob/62f8da293f1c7da18096d7beb9c52f8797712d56/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/fields/ExtractedField.java#L118

Reply:
The tricky part is the condition checks the value is **not** a `long`. Thus, the logic there is that prior to 6.0, we expect a `long`. If it's not, then something's gone wrong. Otherwise, we fall through the last `return` of the method. Pretty confusing, I know.